### DISCUSSION TITLE
Installing extras for Python packages

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#2251] [Everyday usage]
I was wondering if it is/would be technically possible to install packages' extras.

Take `black` as the perfect example.

`black`, the formula, lists the following resources:
* aiohttp
* aiohttp_cors
* async-timeout
* attrs
* chardet
* click
* idna
* multidict
* mypy-extensions
* pathspec
* platformdirs
* regex
* tomli
* typing-extensions
* yarl

On the other hand, "plain" `black` lists the following packages:
```
$ pipgrip --tree black
black (21.9b0)
├── click>=7.1.2 (8.0.1)
├── mypy-extensions>=0.4.3 (0.4.3)
├── pathspec<1,>=0.9.0 (0.9.0)
├── platformdirs>=2 (2.4.0)
├── regex>=2020.1.8 (2021.9.30)
├── tomli<2.0.0,>=0.2.6 (1.2.1)
└── typing-extensions>=3.10.0.0 (3.10.0.2)
```

Currently `black`, the package, offers the following extras:
* "d": ["aiohttp>=3.7.4"]
* "colorama": ["colorama>=0.4.3"]
* "python2": ["typed-ast>=1.4.3"]
* "uvloop": ["uvloop>=0.15.2"]
* "jupyter": ["ipython>=7.8.0", "tokenize-rt>=3.2.0"]

I remember that `brew` depends on `pipgrip` to determine dependencies, and I have found that `pipgrip` is able to list the correct list of dependencies when the `<package_name>[extra]` is passed.

Examples:

<details><summary><strong>black[d]</strong></summary>
<p>

```bash
$ pipgrip --tree "black[d]"
black[d] (21.9b0)
├── aiohttp-cors>=0.4.0 (0.7.0)
│   └── aiohttp>=1.1 (3.7.4.post0)
│       ├── async-timeout<4.0,>=3.0 (3.0.1)
│       ├── attrs>=17.3.0 (21.2.0)
│       ├── chardet<5.0,>=2.0 (4.0.0)
│       ├── multidict<7.0,>=4.5 (5.2.0)
│       ├── typing-extensions>=3.6.5 (3.10.0.2)
│       └── yarl<2.0,>=1.0 (1.6.3)
│           ├── idna>=2.0 (3.2)
│           └── multidict>=4.0 (5.2.0)
├── aiohttp>=3.6.0 (3.7.4.post0)
│   ├── async-timeout<4.0,>=3.0 (3.0.1)
│   ├── attrs>=17.3.0 (21.2.0)
│   ├── chardet<5.0,>=2.0 (4.0.0)
│   ├── multidict<7.0,>=4.5 (5.2.0)
│   ├── typing-extensions>=3.6.5 (3.10.0.2)
│   └── yarl<2.0,>=1.0 (1.6.3)
│       ├── idna>=2.0 (3.2)
│       └── multidict>=4.0 (5.2.0)
├── click>=7.1.2 (8.0.1)
├── mypy-extensions>=0.4.3 (0.4.3)
├── pathspec<1,>=0.9.0 (0.9.0)
├── platformdirs>=2 (2.4.0)
├── regex>=2020.1.8 (2021.9.30)
├── tomli<2.0.0,>=0.2.6 (1.2.1)
└── typing-extensions>=3.10.0.0 (3.10.0.2)
```

</p>
</details>

<details><summary><strong>black[jupyter]</strong></summary>
<p>

```bash
$ pipgrip --tree "black[jupyter]"
black[jupyter] (21.9b0)
├── click>=7.1.2 (8.0.1)
├── ipython>=7.8.0 (7.28.0)
│   ├── appnope (0.1.2)
│   ├── backcall (0.2.0)
│   ├── decorator (5.1.0)
│   ├── jedi>=0.16 (0.18.0)
│   │   └── parso<0.9.0,>=0.8.0 (0.8.2)
│   ├── matplotlib-inline (0.1.3)
│   │   └── traitlets (5.1.0)
│   ├── pexpect>4.3 (4.8.0)
│   │   └── ptyprocess>=0.5 (0.7.0)
│   ├── pickleshare (0.7.5)
│   ├── prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 (3.0.20)
│   │   └── wcwidth (0.2.5)
│   ├── pygments (2.10.0)
│   ├── setuptools>=18.5 (58.2.0)
│   └── traitlets>=4.2 (5.1.0)
├── mypy-extensions>=0.4.3 (0.4.3)
├── pathspec<1,>=0.9.0 (0.9.0)
├── platformdirs>=2 (2.4.0)
├── regex>=2020.1.8 (2021.9.30)
├── tokenize-rt>=3.2.0 (4.1.0)
├── tomli<2.0.0,>=0.2.6 (1.2.1)
└── typing-extensions>=3.10.0.0 (3.10.0.2)
```

</p>
</details>

<details><summary><strong>black[python2]</strong></summary>
<p>

```bash
$ pipgrip --tree "black[python2]"
black[python2] (21.9b0)
├── click>=7.1.2 (8.0.1)
├── mypy-extensions>=0.4.3 (0.4.3)
├── pathspec<1,>=0.9.0 (0.9.0)
├── platformdirs>=2 (2.4.0)
├── regex>=2020.1.8 (2021.9.30)
├── tomli<2.0.0,>=0.2.6 (1.2.1)
├── typed-ast>=1.4.2 (1.4.3)
└── typing-extensions>=3.10.0.0 (3.10.0.2)
```

</p>
</details>

So just like `black` the formula is equivalent to `pip`-installing `black[d]`, I was wondering if new formulae could/should be created in order to give the users the option to install `black[python2]` or `black[jupyter]` and other extras for other packages, or if all possible "extras" should be added to the `black` formula and other formulae.

**Comment by reegnz** [comment]
I ran into a similar issue with `jinja2-cli`. It has `jinja2-cli[yaml]`, `jinja2-cli[toml]` and `jinja2-cli[xml]` extras packages.

I could update the `jinja2-cli` package with:
```
brew update-python-resources jinja2-cli --extra-packages pyyaml,toml,xmltodict|
```

But I cannot use the original `jinja2-cli[yaml]`, `jinja2-cli[toml]` and `jinja2-cli[xml]` extras packages.
Also the `bump-formula-pr` command just nullifies my added dependencies.

Would be great if a Formula for python could take a list of extras that are also to be installed next to the base package.

Right now my workaround is to not use `jinja2-cli` from homebrew (because I want to use it with yaml) but install it with `pipx` instead.

**Reply by alebcay** [reply]
Making edits to [`pypi_formula_mappings.json`](https://github.com/Homebrew/homebrew-core/blob/master/pypi_formula_mappings.json) should be reflected in `brew update-python-resources` and `brew bump-formula-pr` if additional packages should be included with a particular formula.

_Whether_ these extra packages should be included in the `jinja2-cli` formula is another topic of discussion. In `homebrew/core`, there are no optional dependencies - either they're included as dependencies or not (or they can live in a separate formula, if appropriate).

**Reply by reegnz** [reply]
yupp, I discovered that and did a PR after I wrote that comment. We'll see about the discussion.
